components/micro-gateway-core/src/main/ballerina/src/gateway/handlers/mutualSSL_handler.bal

```diff
@@ -39,30 +56,107 @@ public type MutualSSLHandler object {
+                                    "API is not associated with the certificate");
+                                    setErrorMessageToInvocationContext(API_AUTH_INVALID_CREDENTIALS);
+                                    return false;
+                                } else {
+                                    printDebug(KEY_AUTHN_FILTER, "MutualSSL handshake status: PASSED");
+                                    doMTSLFilterRequest(req, invocationContext);
+                                    return true;
+                                }
+                            } else {
+                                handle|error aliasAFromHeaderCert = getAliasAFromHeaderCert(headerValue);
+                                if (aliasAFromHeaderCert is error) {
+                                    setErrorMessageToInvocationContext(API_AUTH_GENERAL_ERROR);
+                                    return prepareAuthenticationError("Unclassified Authentication Failure");
+                                }
+                                if (aliasAFromHeaderCert is handle) {
+                                    boolean isExistAlias = isExistApiAlias(apiVersion, apiName, aliasAFromHeaderCert.toString(),
+                                    self.apiCertificateList);
+                                    if (!isExistAlias || aliasAFromHeaderCert.toString() == "") {
+                                        printDebug(KEY_AUTHN_FILTER, "Mutual SSL authentication failure. API is not associated " +
+                                        "with the certificate");
+                                        self.gatewayCache.addMutualSslCertificateCache(cacheKey, false);
+                                        setErrorMessageToInvocationContext(API_AUTH_INVALID_CREDENTIALS);
+                                        return false;
+                                    } else {
+                                        printDebug(KEY_AUTHN_FILTER, "MutualSSL handshake status: PASSED");
+                                        doMTSLFilterRequest(req, invocationContext);
+                                        self.gatewayCache.addMutualSslCertificateCache(cacheKey, true);
+                                        return true;
+                                    }
+                                }
+
+                            }
+                        } else {
+                            printDebug(KEY_AUTHN_FILTER, "Header has empty value sent by the payload");
+                            setErrorMessageToInvocationContext(API_AUTH_INVALID_CREDENTIALS);
+                            return false;
+                        }
+                    }
+                }
+                string? cert = req.mutualSslHandshake["base64EncodedCert"];
+                var cacheKey = cert.toString() + apiName + apiVersion;
+                var isExistCertCache = self.gatewayCache.retrieveFromMutualSslCertificateCache(cacheKey);
+                if (isExistCertCache is boolean) {
```

Comment:
Why do we expect cert to be always present in the cache. If it is not in the cache shouldn't we go and check the cert in the trust store